mypy/checkexpr.py

```diff
@@ -2122,15 +2098,110 @@ def infer_function_type_arguments(
+                #  solve(outer + inner) and solve(inner + outer) may yield different results.
+                #  we need to use outer first.
+                joint_solution = solve_constraints(
+                    callee_type.variables,
+                    outer_constraints + inner_constraints,
+                    strict=self.chk.in_checked_function(),
+                    allow_polymorphic=False,
+                )
+                joint_args = [
+                    None if has_uninhabited_component(arg) or has_erased_component(arg) else arg
+                    for arg in joint_solution[0]
+                ]
+                joint_solution = (joint_args, joint_solution[1])
+                joint_callee = self.apply_generic_arguments(
+                    callee_type, joint_solution[0], context, skip_unsatisfied=True
+                )
+                joint_ret_type = get_proper_type(joint_callee.ret_type)
+
+                if (  # determine which solution to take
+                    # joint constraints failed to produce a complete solution
+                    None in joint_solution[0]
+                    # If the outer solution is more concrete than the joint solution, prefer the outer solution.
+                    or is_subtype(outer_ret_type, joint_ret_type)
+                    or (  # HACK to fix testLiteralAndGenericWithUnion
+                        isinstance(outer_ret_type, UnionType)
+                        and any(is_subtype(val, joint_ret_type) for val in outer_ret_type.items)
+                    )
+                ):
+                    use_joint = False
+                else:
+                    use_joint = True
+
+                if use_joint:
+                    inferred_args = joint_solution[0]
+                else:
+                    # If we cannot use the joint solution, fallback to outer_solution
+                    inferred_args = outer_solution[0]
+                    # Don't show errors after we have only used the outer context for inference.
+                    # We will use argument context to infer more variables.
+                    callee_type = self.apply_generic_arguments(
+                        callee_type, inferred_args, context, skip_unsatisfied=True
+                    )
+                    if need_refresh:
+                        # Argument kinds etc. may have changed due to
+                        # ParamSpec or TypeVarTuple variables being replaced with an arbitrary
+                        # number of arguments; recalculate actual-to-formal map
+                        formal_to_actual = map_actuals_to_formals(
+                            arg_kinds,
+                            arg_names,
+                            callee_type.arg_kinds,
+                            callee_type.arg_names,
+                            lambda i: self.accept(args[i]),
+                        )
+
+                    # ??? QUESTION: Do we need to recompute arg_types and pass1_args here???
```

Comment:
It seems to work both with/without recomputing `arg_types` here, but I am not sure.